core/lib/zksync_core/src/sync_layer/gossip/storage/mod.rs

```diff
@@ -67,23 +70,117 @@ impl CursorWithCachedBlock {
 #[derive(Debug)]
 pub(super) struct PostgresBlockStorage {
     pool: ConnectionPool,
+    first_block_number: MiniblockNumber,
     actions: ActionQueueSender,
     block_sender: watch::Sender<BlockNumber>,
     cursor: Mutex<CursorWithCachedBlock>,
 }
 
 impl PostgresBlockStorage {
     /// Creates a new storage handle. `pool` should have multiple connections to work efficiently.
-    pub fn new(pool: ConnectionPool, actions: ActionQueueSender, cursor: FetcherCursor) -> Self {
+    pub async fn new(
+        ctx: &ctx::Ctx,
+        pool: ConnectionPool,
+        actions: ActionQueueSender,
+        cursor: FetcherCursor,
+        genesis_block: &FinalBlock,
```

Comment:
AFAIU, this approach won't work for initializing the main node storage; we'll have the chicken-and-egg problem. (Where do we get the genesis block params from? Either we'll need to sign blocks retroactively, or we'll just have the block number.) OTOH, I have some doubts that `PostgresBlockStorage` could be used for the main node as-is (of course, most of it could be reused, but it may require some variability depending on the node role).